src/cached_policy.hpp

```diff
@@ -50,6 +52,14 @@ class CachedPolicy {
     int cache_hits() const;
     int cache_misses() const;
 
+    // Returns the underlying BatchedModel if this policy wraps one, nullptr otherwise
+    std::shared_ptr<BatchedModel> get_batched_model() const {
```

Comment:
Yeah this is pretty evil. Basically using a dynamic cast + friend to break two layers of abstraction. This class is just a caching layer, it takes an evaluation function and turns it into another evaluation function with an LRU cache. Having a function on that which is like "oh yeah if the eval function that I'm caching happens to be an ML model, here's ownership access to that model that I got through my random friend access to that class" is pretty wild.